app/models/bookmark.rb

```diff
@@ -8,12 +8,99 @@ class Bookmark < ActiveRecord::Base
     "reminder_type" # TODO 2021-04-01: remove
   ]
 
+  cattr_accessor :registered_bookmarkables
+  self.registered_bookmarkables = []
+
+  def self.register_bookmarkable(
+    model:, serializer:, list_query:, search_query:, preload_associations: []
+  )
+    Bookmark.registered_bookmarkables << Bookmarkable.new(
+      model: model,
+      serializer: serializer,
+      list_query: list_query,
+      search_query: search_query,
+      preload_associations: preload_associations
+    )
+  end
+
+  ##
+  # This is called when the app loads, similar to AdminDashboardData.reset_problem_checks,
+  # so the default Post and Topic bookmarkables are registered on
+  # boot.
+  #
+  # This method also can be used in testing to reset bookmarkables between
+  # tests. It will also fire multiple times in development mode because
+  # classes are not cached.
+  def self.reset_bookmarkables
+    self.registered_bookmarkables = []
+    Bookmark.register_bookmarkable(
+      model: Post,
+      serializer: UserPostBookmarkSerializer,
+      list_query: lambda do |user, guardian|
+        topics = Topic.listable_topics.secured(guardian)
+        pms = Topic.private_messages_for_user(user)
+        post_bookmarks = user
+          .bookmarks
+          .joins("INNER JOIN posts ON posts.id = bookmarks.bookmarkable_id AND bookmarks.bookmarkable_type = 'Post'")
+          .joins("LEFT JOIN topics ON topics.id = posts.topic_id")
+          .joins("LEFT JOIN topic_users ON topic_users.topic_id = topics.id")
+          .where("topic_users.user_id = ?", user.id)
+          .where(bookmarkable_type: "Post")
+        guardian.filter_allowed_categories(
+          post_bookmarks.merge(topics.or(pms)).merge(Post.secured(guardian))
+        )
+      end,
+      search_query: lambda do |bookmarks, query, ts_query, &bookmarkable_search|
+        bookmarkable_search.call(
```

Comment:
This is kind of weird and I wanted to get your thoughts on this @tgxworld . I need the block because we always need to also search the `bookmarks.name` column and I don't want to leave it up to the bookmarkable being registered to do that. Here is where we are calling `search_query` from the `Bookmarkable` class:

```ruby
  def perform_search_query(bookmarks, query, ts_query)
    search_query.call(bookmarks, query, ts_query) do |bookmarks_joined, where_sql|
      bookmarks_joined.where("#{where_sql} OR bookmarks.name ILIKE :q", q: query)
    end
  end

```

Does this look sane to you?